rpc/client_test.go

```diff
@@ -2,150 +2,160 @@ package rpc
-		t.Errorf("Not on mainnet node: %v\n", err)
-	}
-
-	_, err = client.Syncing(context.Background())
-	if err != nil {
-		t.Errorf("Could not retrieve sync information: %v\n", err)
+var (
+	// set the environment for the test, default: mock
+	testEnv = "mock"
+
+	// testConfigurations are predefined test configurations
+	testConfigurations = map[string]testConfiguration{
+		// Requires a Mainnet StarkNet JSON-RPC compliant node (e.g. pathfinder)
+		// (ref: https://github.com/eqlabs/pathfinder)
+		"mainnet": {
+			base: "http://localhost:9545",
+		},
+		// Requires a Testnet StarkNet JSON-RPC compliant node (e.g. pathfinder)
+		// (ref: https://github.com/eqlabs/pathfinder)
+		"testnet": {
+			base: "http://localhost:9545",
+		},
+		// Requires a Devnet configuration running locally
+		// (ref: https://github.com/Shard-Labs/starknet-devnet)
+		"devnet": {
+			base: "http://localhost:5050/rpc",
+		},
+		// Used with a mock as a standard configuration, see `mock_test.go``
+		"mock": {},
 	}
-}
+)
 
-func TestBlock(t *testing.T) {
-	num, err := client.BlockNumber(context.Background())
-	if err != nil {
-		t.Errorf("Could not retrieve block: %v\n", err)
-	}
+// TestMain is used to trigger the tests and, in that case, check for the environment to use.
+func TestMain(m *testing.M) {
+	flag.StringVar(&testEnv, "env", "mock", "set the test environment")
+	flag.Parse()
 
-	block, err := client.BlockByNumber(context.Background(), num, "FULL_TXN_AND_RECEIPTS")
-	if err != nil {
-		t.Errorf("Could not retrieve block: %v\n", err)
-	}
-	if block.Transactions[0].TransactionReceipt.Status == "" {
-		t.Errorf("Could not retrieve transaction receipts: %v\n", block.Transactions[0])
-	}
+	os.Exit(m.Run())
+}
 
-	_, err = client.BlockByNumber(context.Background(), num, "FULL_TXNS")
-	if err != nil {
-		t.Errorf("Could not retrieve block: %v\n", err)
+// beforeEach checks the configuration and initializes it before running the script
+func beforeEach(t *testing.T) *testConfiguration {
+	godotenv.Load(fmt.Sprintf(".env.%s", testEnv), ".env")
+	testConfig, ok := testConfigurations[testEnv]
+	if !ok {
+		t.Fatal("env supports mock, testnet, mainnet or devnet")
 	}
-
-	block, err = client.BlockByHash(context.Background(), block.BlockHash, "FULL_TXN_AND_RECEIPTS")
-	if err != nil {
-		t.Errorf("Could not retrieve block: %v\n", err)
+	if testEnv == "mock" {
+		testConfig.client = &Client{
+			c: &rpcMock{},
+		}
+		return &testConfig
 	}
-	if block.Transactions[0].TransactionReceipt.Status == "" {
-		t.Errorf("Could not retrieve transaction receipts: %v\n", block.Transactions[0])
+	base := os.Getenv("INTEGRATION_BASE")
+	if base != "" {
+		testConfig.base = base
 	}
-
-	_, err = client.BlockByHash(context.Background(), block.BlockHash, "FULL_TXNS")
+	client, err := DialContext(context.Background(), testConfig.base)
 	if err != nil {
-		t.Errorf("Could not retrieve block: %v\n", err)
+		t.Fatal("connect should succeed, instead:", err)
 	}
+	testConfig.client = client
```

Comment:
Thoughts on encapsulating the teardown here rather than in each caller?
```suggestion
	testConfig.client = client
	t.Cleanup(func() {
		testConfig.client.Close()
	})
```

Reply:
I did not know that. That makes a lot of sense. I'll push the change.